stdlib/src/builtin/builtin_slice.mojo

```diff
@@ -158,11 +146,31 @@ struct Slice(Sized, Stringable, EqualityComparable):
         Returns:
             The slice index.
         """
-        return self.start + idx * self.step
+        return self.start.value() + idx * self.step
 
     @always_inline("nodebug")
     fn _has_end(self) -> Bool:
-        return self.end != _int_max_value()
+        return self.end.__bool__()
+
+    fn adjust(inout self, src_len: Int):
+        """Adjusts span.
+
+        Uses target container length to adjust negative or None indices.
+
+        Args:
+            src_len: The length of the target container.
+        """
+        var negative_step = self.step < 0
+
+        if not self.start:
+            self.start = 0 if not negative_step else src_len - 1
+        elif self.start.value() < 0:
+            self.start = self.start.value() + src_len
```

Comment:
Need bounds checks.  

- Given a negative start and positive step, Python bounds `start >= 0` 
- Given a negative start and negative step,  Python bounds `start >= -1`
- Given a positive start and negative step, Python bounds `start <= src_len - 1`